File: app/src/main/java/com/tangem/data/network/TezosApi.java
package com.tangem.data.network;

import com.tangem.data.network.model.TezosAccountResponse;
import com.tangem.data.network.model.TezosForgeBody;
import com.tangem.data.network.model.TezosHeaderResponse;
import com.tangem.data.network.model.TezosPreapplyBody;

import java.util.List;

import io.reactivex.Single;
import retrofit2.Call;
import retrofit2.http.Body;
import retrofit2.http.GET;
import retrofit2.http.POST;
import retrofit2.http.Path;

public interface TezosApi {
    @GET(ServerApiTezos.TEZOS_ADDRESS)
    Single<TezosAccountResponse> getAccount(@Path("address") String address);

    @GET(ServerApiTezos.TEZOS_HEADER)
    Call<TezosHeaderResponse> getHeader();

    @GET(ServerApiTezos.TEZOS_MANAGER_KEY)
    Single<String> getManagerKey(@Path("address") String address);

    @POST(ServerApiTezos.TEZOS_FORGE_OPERATIONS)
    Call<String> forgeOperations(@Body TezosForgeBody tezosForgeBody);

    @POST(ServerApiTezos.TEZOS_PREAPPLY_OPERATIONS)
    Call<Void> preapplyOperations(@Body List<TezosPreapplyBody> tezosPreapplyBodyList);

    @POST(ServerApiTezos.TEZOS_INJECT_OPERATIONS)
    Single<Object> injectOperations(@Body String txForSend);
}

File: blockchain/src/main/java/com/tangem/blockchain/blockchains/tezos/network/TezosApi.kt
package com.tangem.blockchain.blockchains.tezos.network

import retrofit2.http.Body
import retrofit2.http.GET
import retrofit2.http.POST
import retrofit2.http.Path

interface TezosApi {
    @GET("chains/main/blocks/head/context/contracts/{address}")
    suspend fun getAddressData(@Path("address") address: String): TezosAddressResponse

    @GET("chains/main/blocks/head/header")
    suspend fun getHeader(): TezosHeaderResponse

    @GET("chains/main/blocks/head/context/contracts/{address}/manager_key")
    suspend fun getManagerKey(@Path("address") address: String): String

    @POST("chains/main/blocks/head/helpers/forge/operations")
    suspend fun forgeOperations(@Body tezosForgeBody: TezosForgeBody): String

    @POST("chains/main/blocks/head/helpers/preapply/operations")
    suspend fun preapplyOperations(@Body tezosPreapplyBodyList: List<TezosPreapplyBody>)

    @POST("injection/operation")
    suspend fun sendTransaction(@Body transaction: String)
}

data class TezosForgeBody(
        val branch: String,
        val contents: List<TezosOperationContent>
)

data class TezosOperationContent(
        val kind: String,
        val source: String,
        val fee: String,
        val counter: String,
        val gas_limit: String,
        val storage_limit: String,
        val public_key: String? = null,
        val destination: String? = null,
        val amount: String? = null
)

data class TezosPreapplyBody(
        val protocol: String,
        val branch: String,
        val contents: List<TezosOperationContent>,
        val signature: String
)